scale-decode/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+// try to convert into the desired output size, failing if this isn't possible:
+assert_decodes_to(123u64, 123u8);
+assert_decodes_to(123u8, 123u64);
+
+// Compact decoding is also handled "under the hood" by DecodeAsType, so no "compact"
+// annotations are needed on values.
+assert_decodes_to(codec::Compact(123u64), 123u64);
+
+// Enum variants are lined up by variant name, so no explicit "index" annotation are
+// needed either; DecodeAsType will take care of it.
+#[derive(Encode, TypeInfo)]
+enum Foo {
+    #[codec(index = 10)]
+    Something(u64),
+}
+#[derive(DecodeAsType, PartialEq, Debug)]
+enum FooTarget {
+    Something(u128),
+}
+assert_decodes_to(Foo::Something(123), FooTarget::Something(123));
+
+// DecodeAsType will skip annotated fields and not look for them in the encoded bytes.
+// #[codec(skip)] and #[decode_as_type(skip)] both work.
+#[derive(Encode, TypeInfo)]
+struct Bar {
+    a: bool,
+}
+#[derive(DecodeAsType, PartialEq, Debug)]
+struct BarTarget {
+    a: bool,
+    #[decode_as_type(skip)]
+    b: String,
+}
+assert_decodes_to(
+    Bar { a: true },
+    BarTarget { a: true, b: String::new() },
+);
+
+// DecodeAsType impls will generally skip through any newtype wrappers.
+#[derive(DecodeAsType, Encode, TypeInfo, PartialEq, Debug)]
+struct Wrapper {
+    value: u64
+}
+assert_decodes_to(
+    (Wrapper { value: 123 },),
+    123u64
+);
+assert_decodes_to(
+    123u64,
+    (123,)
+);
+
+// Things like arrays and sequences are generally interchangeable despite the
+// encoding format being slightly different:
+assert_decodes_to([1u8,2,3,4,5], vec![1u64,2,3,4,5]);
+assert_decodes_to(vec![1u64,2,3,4,5], [1u8,2,3,4,5]);
+```
+
+If this high level interface isn't suitable, you can implement your own [`visitor::Visitor`]'s. These can support zero-copy decoding
+(unlike the higher level [`DecodeAsType`] interface), and generally the Visitor construction and execution is zero alloc, allowing
+for efficient type based decoding.
+*/
+#![deny(missing_docs)]
+
+mod impls;
+mod utils;
+
+pub mod error;
+pub mod visitor;
+
+pub use crate::error::Error;
+pub use visitor::Visitor;
+
+#[cfg(feature = "derive")]
+pub use scale_decode_derive::DecodeAsType;
+
+// Used in trait definitions.
+pub use scale_info::PortableRegistry;
+/// A description of a single field in a tuple or struct type.
+pub type PortableField = scale_info::Field<scale_info::form::PortableForm>;
+/// A type ID used to represent tuple fields.
+pub type PortableFieldId = scale_info::interner::UntrackedSymbol<std::any::TypeId>;
+
+/// This trait is implemented for any type `T` where `T` implements [`IntoVisitor`] and the errors returned
+/// from this [`Visitor`] can be converted into [`Error`]. It's essentially a convenience wrapper around
+/// [`visitor::decode_with_visitor`] that mirrors `scale-encode`'s `EncodeAsType`.
+pub trait DecodeAsType: Sized {
+    /// Given some input bytes, a `type_id`, type registry and context, attempt to decode said bytes into
```

Comment:
nit: Is this the `context` that was previously used to build the error info?

Reply:
Well spotted! Yup, this context is no longer passed in so I updated the comment :)